docs/src/math.md

```diff
@@ -0,0 +1,112 @@
+Such a type we will call a **Primal Type**.
+
+Roughly speaking, the **Differential Type** is a type that represents the difference between two primal type value.
+For a given Primal type there will often be multiple
+valid Differential types.
+For example, for the Primal Type of `DateTime`, the valid Differential Types include: `Millisecond`, `Second`, `Hour`, `Day` etc.
+
+!!! note
+    There will always be one (the **Zero**).
+    If there is only that one then one normally says that the type is not differentiable.
+    For example booleans, and integers except when they are being uses as special cases of real numbers; as a computational optimization.
+
+
+### Direct Differential Types
+For purposes of definitions, we introduct the notion of a direct differntial.
+This is a much weaker requirement than we actually need, but will make the defintions that follow easier.
+A direct differntial type for a given primal type is something that can be added to an instance primal type to get another instance of the primal type.
+
+Formally speaking:
+For a given Primal Type ``\mathcal P``,
+qnd a type ``\mathcal D``.
+If for all values ``d::\mathcal D``
+and for all values ``p::\mathcal P``
+there exists a ``q::\mathcal P`` such that
+``p + d = d + p = q``.
+Then we say that  ``\mathcal D`` is a direct-differential type for the primal type ``\mathcal P``.
+Which we write as ``\mathcal D \triangleleft \mathcal P``.
+
+The valid differntial extends the requirements of a direct differnetial to also require a form of closed-ness under addition.
+
+### Valid Differential Types
+As a stronger condition, we define a valid differential type for a given primal type.
+This is what we actually require.
+Consider a Primal Type ``\mathcal P``.
+
+Consider some type ``\mathcal D \triangleleft P``.
+
+If there exists a type-union ``\mathbb D \triangleleft \mathcal P``, with ``\mathcal D <: \mathbb D``.
+(i.e. a type union of direct differential types for ``\mathcal P`` exist, and ``\mathcal D`` is part of that union);
+and for all ``d_x :: \mathcal D`` and forall ``d_u :: \mathbb D``,
+there exists a ``d_s :: \mathbb D``
+such that ``d_x + d_u = d_u + d_x = d_s``,
+then we say that ``\mathcal D`` is a valid differential type for ``\mathcal P`.
+And we write this as ``\mathcal D \blacktriangleleft \mathcal P``.
+
+Note: in this case it is also true that every other type in the type-union ``\mathbb D`` and ``\mathbb D`` itself are also valid differential types for ``\mathcal P``.
+
+
+The informal version of this if a direct-differential type are closed under addition to always give a instance of a direct differential type (potentiall the same one), then it is a valid differential type.
+The important take away is you can add instances of all valid differential types for ``\mathcal P``,
+and know you will always get an instance of valid differential type back.
+
+
+-----
+
+### REWRITE: Differential Type
```

Comment:
Could we simplify even further by saying something like the following:

A type ``\mathcal D`` is a valid differential type for primal type ``\mathcal P`` if
- addition between a primal in ``\mathcal P`` and differential in ``\mathcal D`` yields another primal in ``\mathcal P``
- addition of two differentials in ``\mathcal D`` produces another differential in ``\mathcal D``. We say that ``\mathcal D`` is closed under addition.

Reply:
We need the ability to add 2 differentials of different types and come back with another differential, potentially of a 3rd type.
Common example would be `[1.0,2.0,3.0] + Zero() = [1.0,2.0,3.0]`

another isi
`Composite{Foo, NamedTuple{(:x,),Tuple{Float64}})(x=2.0) + Composite{Foo, NamedTuple{(:y,),Tuple{Float64}})(y=3.0) = Composite{Foo, NamedTuple{(:x,:y),Tuple{Float64,Float64}})(x=2.0, y=3.0)`